readthedocs/core/views/serve.py

```diff
@@ -269,3 +272,105 @@ def robots_txt(request, project):
         return HttpResponse(open(fullpath).read(), content_type='text/plain')
 
     return HttpResponse('User-agent: *\nAllow: /\n', content_type='text/plain')
+
+
+@map_project_slug
+# TODO: make this cache dependent on the project's slug
+@cache_page(60 * 60 * 24 * 3)  # 3 days
+def sitemap_xml(request, project):
+    """
+    Generate and serve a ``sitemap.xml`` for a particular ``project``.
+
+    The sitemap is generated from all the ``active`` and public versions of
+    ``project``. These versions are sorted by using semantic versioning
+    prepending ``latest`` and ``stable`` (if they are enabled) at the beginning.
+
+    Following this order, the versions are assigned priorities and change
+    frequency. Starting from 1 and decreasing by 0.1 for priorities and starting
+    from daily, weekly to monthly for change frequency.
+
+    If the project is private, the view raises ``Http404``. On the other hand,
+    if the project is public but a version is private, this one is not included
+    in the sitemap.
+
+    :param request: Django request object
+    :param project: Project instance to generate the sitemap
+
+    :returns: response with the ``sitemap.xml`` template rendered
+
+    :rtype: django.http.HttpResponse
+    """
+    def priorities_generator():
+        """
+        Generator returning ``priority`` needed by sitemap.xml.
+
+        It generates values from 1 to 0.1 by decreasing in 0.1 on each
+        iteration. After 0.1 is reached, it will keep returning 0.1.
+        """
+        priorities = [1, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2]
+        yield from itertools.chain(priorities, itertools.repeat(0.1))
+
+    def changefreqs_generator():
+        """
+        Generator returning ``changefreq`` needed by sitemap.xml.
+
+        It returns ``daily`` on first iteration, then ``weekly`` and then it
+        will return always ``monthly``.
+
+        We are using ``monthly`` as last value because ``never`` is too
+        aggressive. If the tag is removed and a branch is created with the same
+        name, we will want bots to revisit this.
+        """
+        changefreqs = ['daily', 'weekly']
+        yield from itertools.chain(changefreqs, itertools.repeat('monthly'))
+
+    if project.privacy_level == constants.PRIVATE:
+        raise Http404
+
+    sorted_versions = sort_version_aware(
+        project.versions.filter(
+            active=True,
+            privacy_level=constants.PUBLIC,
+        ),
+    )
+
+    versions = []
+    for version, priority, changefreq in zip(
+            sorted_versions, priorities_generator(), changefreqs_generator()):
```

Comment:
If I'm reading this right:

- `latest` will always have priority=1 and changefreq=daily
- `stable` will always have priority=0.9 and changefreq=weekly
- other versions will have decreasing priorities and changefreq=monthly

Is that right?

Wouldn't it be better to just guess at a priority and changefrequency based on the last build date? If there is no last build date (version was never built), we don't include the version.

Reply:
You are reading it right, yes.

> Wouldn't it be better to just guess at a priority and changefrequency based on the last build date?

Our idea behind this was that you (as author of the project) want to point your readers to the latest version published first. That's why the priority works like that: latest, stable, v1.5, v1.4, etc in that order... I think that build date is not associated with priority.

Regarding changefreq, I have a similar opinion: I expect last versions to change more frequently than v0.1.

I wouldn't complicate the logic for this.

> If there is no last build date (version was never built), we don't include the version.

I'm including only `active` versions, as we do on the flyout also. This could change once more states be implemented: https://github.com/rtfd/readthedocs.org/issues/4001#issuecomment-437549598